src/emitEventsInBatches.js

```diff
@@ -0,0 +1,55 @@
+const chunk = require('lodash.chunk');
+const omit = require('lodash.omit');
+const { enrichMeta, partitionKey } = require('./enrich')
+
+const MAX_RECORDS = 500; 
+
+const enrichRecords = (events, config) => {
+  return events.map(event => {
+    const enrichedEvent = enrichMeta(event, config.appName, config.ipv4);
+    return {
+      Data: JSON.stringify(enrichedEvent),
+      PartitionKey: config.partitionKey || partitionKey(enrichedEvent),
+    }
+  })
+};
+
+const emitEvents = async (kinesis, records, config, retries) => {
+  const params = {
+    Records: records,
+    StreamName: config.kinesisStream.resource
+  };
+
+  try {
+    const { FailedRecordCount, Records } = await kinesis.putRecords(params).promise();
+
+    if(FailedRecordCount !== 0) {
+      const failedEvents = [];
+      Records.forEach((failedRecord, index) => {
+        if (failedRecord.ErrorCode) {
+          failedEvents.push({
+            failedRecord: records[index],
+            failedEvent: JSON.parse(records[index].Data),
+            failureMessage: `${failedRecord.ErrorCode}: ${failedRecord.ErrorMessage}`
+          });
+        }
+      });
+
+      if (retries === 0) throw failedEvents.map(failed => omit(failed, ['failedRecord'])); 
+
+      return await emitEvents(kinesis, failedEvents.map(failed => failed.failedRecord), config, retries - 1);
+    }
+  } catch(error) {
+    if (retries === 0) throw error
+    return await emitEvents(kinesis, records, config, retries - 1);
+  }
+};
+
+module.exports = (kinesis, events, config) => {
+  const retries = config.maxRetries || 0;
+  const emitEventsPromises = chunk(events, MAX_RECORDS).map(async (chunkedEvents) => {
```

Comment:
I don't think you can make this one an `async` function, as `map` is not made to await.

Instead I think you can keep what you had before:
```javascript
return emitEvents(kinesis, enrichedRecords, config, retries);
```
So now the `emitEventsPromises` array contains async operations in progress, which should be turned into Promises, which should work for `Promise.allSettled`.

I tried in the browser doing this:
```javascript
// Declaring a test async funciton
let asyncFunc = async () => await new Promise(resolve => {
  window.setTimeout(() => resolve("resolved"), 500);
});
> undefined
// Awaiting it works as expected
await asyncFunc()
> 'resolved'
// Calling it without awaiting returns a pending promise, which is perfect for your usecase
asyncFunc()
> Promise {<pending>}
```

Sorry I might have added doubts with other comments, mixing async/await and Promises is also somewhat new to me 😅

Reply:
hehe 😅 reverted the other function as well 👍